galley/pkg/config/analysis/analyzers/virtualservice/regexes.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package virtualservice
+
+import (
+	"regexp"
+
+	"istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+// RegexAnalyzer checks all regexes in a virtual service
+type RegexAnalyzer struct{}
+
+var _ analysis.Analyzer = &RegexAnalyzer{}
+
+// Metadata implements Analyzer
+func (a *RegexAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name:        "virtualservice.RegexAnalyzer",
+		Description: "Checks regex syntax",
+		Inputs: collection.Names{
+			collections.IstioNetworkingV1Alpha3Virtualservices.Name(),
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (a *RegexAnalyzer) Analyze(ctx analysis.Context) {
+	ctx.ForEach(collections.IstioNetworkingV1Alpha3Virtualservices.Name(), func(r *resource.Instance) bool {
+		a.analyzeVirtualService(r, ctx)
+		return true
+	})
+}
+
+func (a *RegexAnalyzer) analyzeVirtualService(r *resource.Instance, ctx analysis.Context) {
+
+	vs := r.Message.(*v1alpha3.VirtualService)
+
+	for _, route := range vs.GetHttp() {
+		for _, m := range route.GetMatch() {
+			analyzeStringMatch(r, m.GetUri(), ctx, "uri")
+			analyzeStringMatch(r, m.GetScheme(), ctx, "scheme")
+			analyzeStringMatch(r, m.GetMethod(), ctx, "method")
+			analyzeStringMatch(r, m.GetAuthority(), ctx, "authority")
+			for _, h := range m.GetHeaders() {
+				analyzeStringMatch(r, h, ctx, "header")
+			}
+			for _, qp := range m.GetQueryParams() {
+				analyzeStringMatch(r, qp, ctx, "query param")
+			}
+			for _, wh := range m.GetWithoutHeaders() {
+				analyzeStringMatch(r, wh, ctx, "without header")
+			}
+		}
+		for _, origin := range route.GetCorsPolicy().GetAllowOrigins() {
+			analyzeStringMatch(r, origin, ctx, "CORS policy allow origins")
+		}
+	}
+}
+
+func analyzeStringMatch(r *resource.Instance, sm *v1alpha3.StringMatch, ctx analysis.Context, where string) {
+	re := sm.GetRegex()
+	if re == "" {
+		return
+	}
+
+	_, err := regexp.Compile(re)
```

Comment:
Is this enough? Envoy has its own regular expression engine; I am not familiar if the syntax is compatible.

Reply:
Ok, I read #21717 and see this was already discussed in the issue. They _should_ be the same.

Of course I'd worry about them diverging; even a version update could cause this to behave differently. Still, since this is an analyzer, the user could work around the issue by ignoring the message or suppressing it. The value provided by actually checking regexp syntax probably outweighs the risk.